doc/design/bucketpath.rst

```diff
@@ -0,0 +1,157 @@
+==============================
+Design Document Bucket Path(s)
+==============================
+
+Introduction
+============
+
+- Purpose of the document / Problem description
+- Definitions and Acronyms
+
+Overview
+========
+
+Design Goals
+============
+
+Path Design
+===========
+
+Future Considerations
+=====================
+
+Sudo Code Outline
+=================
+
+.. code-block:: python
+
+    from exasol.bucketfs import PathBuilder
+
+    Path = PathBuilder(credentials)
+
+    udf_path = Path("bfsl://some/local/path/file.tar.gz")
+    http_bucket_path = Path("bfs://127.0.0.1:8888/service/bucket/some/file.tar.gz")
+    https_bucket_path = Path("bfss://127.0.0.1:8888/service/bucket/some/file.tar.gz")
+    chroot_path = Path("bfss://127.0.0.1:8888/service/bucket/some/sub/subsub/file.tar.gz?chroot=/some/sub/")
+
+
+
+.. code-block:: python
+
+    from typing import Protocol
+
+    # interface / protocol or abc
+    # Note: should mimic pathlib.Path so it can easily understood and adopted. 
+    class Pathlike(Protocol):
+
+        @property
+        def name:
+            ...
+
+        @property
+        def suffix:
+            ...
+
+        @property
+        def root:
+            ...
+
+        def as_uri():
+            ...
+
+        def exists():
+            ...
+
+        def is_dir():
+            ...
+
+        def is_file():
+            ...
+
+        def read_text():
+            ...
+
+        def read_bytes():
+            ...
+
+        def write_text():
+            ...
+
+        def write_bytes():
+            ...
+
+        def unlink():
+            ...
+
+        def rmdir():
+            ...
```

Comment:
In Bucketfs they don't but this abstraction, shall provide the means of "Directories" right. 
If we want to be able to delete all bellow that dir I think we should add it, if we don't want to support directory deletion that's fine too. 

Regarding the unlink I am not sure it would help understand `pathlib.Path`  users more easily whats going on, on the other hand it may isn't obvious for non FS people so an `rm` or `delete` may would be better.

Reply:
Yeah, I think, we need to iterate over this and check the exact meaning of these functions.

Reply:
I had a look at unlink and rmdir. Unlink is ok, because it only works on files. However, rmdir only works for empty directories which doesn't exist in bucketfs, because as soon a directory gets empty in the bucketfs it automatically vanish.

Reply:
ok to stay compatible with the API, we can implement rmdir as you suggested. And, yes we probably need something like rmtree